include/modules/dwm.hpp

```diff
@@ -0,0 +1,371 @@
+
+    /**
+     * Get the address to the layout represented by the symbol.
+     */
+    auto find_layout(const string& sym) const -> const dwmipc::Layout*;
+
+    /**
+     * Get the address to the layout represented by the address.
+     */
+    auto find_layout(uintptr_t addr) const -> const dwmipc::Layout*;
+
+    /**
+     * Get the address of the next layout in m_layouts.
+     *
+     * @param layout Address of the current layout
+     * @param wrap True to wrap around the array, false to return the same
+     *   layout if the next layout does not exist.
+     */
+    auto next_layout(const dwmipc::Layout& layout, bool wrap) const -> const dwmipc::Layout*;
+
+    /**
+     * Get the address of the previous layout in m_layouts.
+     *
+     * @param layout Address of the current layout
+     * @param wrap True to wrap around the array, false to return the same
+     *   layout if the next layout does not exist.
+     */
+    auto prev_layout(const dwmipc::Layout& layout, bool wrap) const -> const dwmipc::Layout*;
+
+    /**
+     * Check if the command matches the specified IPC command name and if so,
+     * parse and send the command to dwm
+     *
+     * @param cmd The command string given by dwm_modue::input
+     * @param ipc_cmd The name of dwm IPC command to check for
+     *
+     * @return true if the command matched, was succesfully parsed, and sent to
+     *   dwm, false otherwise
+     */
+    auto check_send_cmd(string cmd, const string& ipc_cmd) -> bool;
+
+    /**
+     * Helper function to build cmd string
+     *
+     * @param ipc_cmd The dwm IPC command name
+     * @param arg The argument to the dwm command
+     */
+    auto static build_cmd(const char* ipc_cmd, const string& arg) -> string;
+
+    /**
+     * Attempt to connect to any disconnected dwm sockets. Catch errors.
+     */
+    auto reconnect_dwm() -> bool;
+
+    /**
+     * If true, enables the click handlers for the tags
+     */
+    bool m_tags_click{true};
+
+    /**
+     * If true, enables the click handlers for the layout label
+     */
+    bool m_layout_click{true};
+
+    /**
+     * If true, scrolling the layout cycle through available layouts
+     */
+    bool m_layout_scroll{true};
+
+    /**
+     * If true, scrolling the layout will wrap around to the beginning
+     */
+    bool m_layout_wrap{true};
+
+    /**
+     * If true, scrolling the layout will cycle layouts in the reverse direction
+     */
+    bool m_layout_reverse{false};
+
+    /**
+     * If true, show floating label
+     */
+    bool m_is_floating{false};
+
+    /**
+     * If the layout symbol is clicked on, it will set the layout represented by
+     * this symbol. The default is monocle mode [M].
+     */
+    string m_secondary_layout_symbol{"[M]"};
+
+    /**
+     * Holds the address to the secondary layout specified by the secondary
+     * layout symbol
+     */
+    const dwmipc::Layout* m_secondary_layout = nullptr;
```

Comment:
Why do we need a pointer for all these variables?

Reply:
They don't necessarily need to be pointers. Most of these variables are updated as events are received and used in other logic for determining what click/scroll handlers should do, etc. I believe they can all be replaced by array indices which would be safer to use. The main difference would be `m_secondary_layout->...` vs `m_layouts->at(m_secondary_layout)`. It's more to simplify that initial array access. Would you prefer to replace these pointers with indices?

Reply:
Hmm, no I think it looks better this way. I was more wondering why we don't just store the instances directly, but if the underlying data is changed by dwmipcpp, this makes sense.

I guess this introduces more concurrency concerns. Does dwmipcpp start any of its own threads? Does it synchronize accesses to these data structures? Acquiring both the update and build locks in the `input` method should probably suffice.

Reply:
Just to clarify, the underlying data isn't changed by dwmipcpp. Most of the pointers are to specific elements of `shared_ptr<vector<...>>` class members which are all initialized in the `dwm_module` constructor. The library creates the shared_pointers to these vectors and returns the shared pointers in the method calls. This data doesn't change; what changes is the module's knowledge of the currently active monitor, currently selected layout, etc. which are kept track of using those pointers for simplicity.

`dwmipcpp` does not start any of its own threads and does not synchronize any access.

Hope this clarifies everything.